### PR TITLE
Mutation testing setup: Phase 1 critical-path coverage + Phase 2 weekly CI on ECMWF infra

### PR DESCRIPTION
**Pull request by tlmquintino**
## Summary

Establishes the mutation testing regime for the tensogram Rust workspace, per the rollout plan in `MUTATION_TESTING.md`. Phase 1 (file-by-file critical-path sweeps) is mostly complete; Phase 2 machinery (sharded weekly sweep on ECMWF self-hosted infrastructure) is in place and ready to fire on the first weekend after merge.

The work is **measurement of existing tests, not new feature code** — no library source paths changed; new tests + per-mutation exemptions document where the existing test suite has genuine coverage and where it has cosmetic gaps.

### New tests (19 across 4 files)
- `rust/tensogram/src/wire.rs`: 1 test (preamble reserved-field round-trip — kills `read_u32_be → 0/1` mutations).
- `rust/tensogram/src/metadata.rs`: 5 tests (CBOR map equality with same keys/different values + vice versa, `canonicalize` Tag recursion, `verify_canonical_value` Array + Tag recursion).
- `rust/tensogram/src/validate/mod.rs`: 6 tests (decompression checks for zstd / encoding-only / filter-only pipelines, corrupt-zstd HashMismatch path, descriptor re-parse fallback).
- `rust/tensogram/tests/decode_verify_hash.rs`: 7 tests (verify_hash on under-PREAMBLE_SIZE buffers, hashed/unhashed `decode_object_from_frame`, mask restoration in `decode_range_from_frame`).

### New configuration + workflows
- `.cargo/mutants.toml` — initial config with skip-globs (remote.rs, doctor, CLI, benchmarks, examples, sz3-sys), 33 documented `exclude_re` patterns covering cosmetic / equivalent / unreachable mutations from steps 1.1–1.7, `--profile release-mutants` selection, `--test-threads=2`.
- `Cargo.toml` — new `[profile.release-mutants]` inheriting release with `codegen-units = 1` (caps rustc's internal codegen fan-out — the layer that was causing dev-laptop power events even at `CARGO_MUTANTS_JOBS=2`).
- `.github/workflows/mutants-weekly.yml` (NEW) — sharded weekly sweep on ECMWF self-hosted runners (`platform-builder-docker-xl`), Saturday + Sunday 09:00 UTC, 16 shards × `max-parallel: 2`, GH-Actions `timeout-minutes: 480` + `timeout 7h` wrapper for safety, structured `\$GITHUB_STEP_SUMMARY` + auto-issue body, `open-issue-on-fail` opens GitHub issue tagged `mutation-testing` / `triage`.
- `.github/workflows/ci.yml` — new `mutants-diff` PR-time job on ECMWF self-hosted runners (non-blocking; flips to required-for-merge after the first clean weekend).
- `.gitignore` — `mutants.out/`, `mutants.out.*/`, `mutants*.log`.

### Documentation
- `docs/src/dev/mutation-testing.md` (NEW, 302 lines, mdbook). Full user-facing developer guide: what mutation testing is, why tensogram uses it, the two regimes (PR-diff + weekly), the seven concurrency clamps, macOS `taskpolicy -b` escape hatch, when to add a test vs. an exemption, how to read `mutants.out/`, version-bump protocol.
- `docs/src/SUMMARY.md` — links the new page under Reference.
- `CONTRIBUTING.md` — Test Structure table mentions mutation testing.
- `plans/TEST.md` — new `## Mutation testing` section (cold-start context for AI Code Agent sessions) + coverage-shape addendum row + edge-case pointer.
- `MUTATION_TESTING.md` (uncommitted at repo root, planning doc) — full rollout plan with seven sections covering goal, scope, two phases, configuration evolution, documentation triangle, acceptance criteria, risks, sequencing checklist, effort estimate.

## Changed

- 33 `exclude_re` entries in `.cargo/mutants.toml`, each with a one-line rationale comment. No mass suppression by category — every exemption is specific (file:line:pattern) and individually justified.
- Concurrency configuration covers seven nested layers, not just `--jobs`:
  1. `CARGO_MUTANTS_JOBS=2` (mutant workers)
  2. `--jobserver-tasks 2` (cargo+rustc)
  3. `codegen-units = 1` (rustc codegen — pinned via `[profile.release-mutants]`)
  4. `--test-threads=2` (test binaries — pinned in mutants.toml)
  5. `CMAKE_BUILD_PARALLEL_LEVEL=1` (cmake-driven build.rs in deps)
  6. `MAKEFLAGS=-j1` (nested make invocations)
  7. `TENSOGRAM_THREADS=1` (library-internal rayon)

  All seven necessary because they multiply, not add — clamping any one alone leaves the others unbounded.

- Format polish: 5 files (3 Python + 2 Rust) brought into `cargo fmt` + `ruff format` compliance. Pre-existing drift unrelated to this branch but caught by the stricter pre-flight.

## Fixed

Nothing. Mutation testing surfaced no actual bugs in shipped tensogram code. The 78 cosmetic survivors triaged across steps 1.1–1.7 are non-bugs (Display string wording, defensive bounds, dead-code arms, parallelism dispatch where input parameters force the branch); the 19 new tests close coverage gaps in already-correct code so future regressions of those code paths get caught.

## Phase 1 status (steps 1.1 – 1.7 complete; step 1.8 deferred)

| Step | File | Mutants | Caught | Survivors triaged | Tests | Exemptions |
|---|---|---:|---:|---|---:|---:|
| 1.1 | `hash.rs` | 32 | 32 | — | 0 | 0 |
| 1.2 | `error.rs` | 3 | 0 | 1 unviable + 2 cosmetic | 0 | 1 |
| 1.3 | `wire.rs` | 68 | 61 | 5 unviable + 2 algebraic | 1 | 2 |
| 1.4 | `dtype.rs` | 7 | 7 | — | 0 | 0 |
| 1.5 | `metadata.rs` | 104 | 68 | 16 unviable + 20 cosmetic | 5 | 3 |
| 1.6 | `validate/integrity.rs` | 61 | 25 | 7 unviable + 29 cosmetic | 6 | 9 |
| 1.7 | `decode.rs` | 83 | 40 | 18 unviable + 25 cosmetic | 7 | 18 |
| **Total 1.1–1.7** | **7 files** | **358** | **233** | **125** | **19** | **33** |

**Step 1.8 `framing.rs` (~551 mutants)**: deferred to Phase 2's weekly sweep per the option-3 decision recorded in `MUTATION_TESTING.md` §7. Two laptop power-down events while running this file under all seven concurrency clamps + `taskpolicy -b` made local execution unsafe. The file is in scope of the workspace-wide weekly sweep without special handling — sharding partitions its mutants alongside everything else.

## Phase 2 status

Workflow file committed and ready. Goes live the first weekend after merge:
- Saturday + Sunday 09:00 UTC
- 16 shards × `max-parallel: 2` ≈ ~4 hr per weekend day
- ECMWF self-hosted runners; **zero GitHub-Actions minutes for the sweep itself** (only the small `open-issue-on-fail` job uses ubuntu-latest, ~10 sec per failed weekend)

## Pre-flight verified

- `cargo fmt --check` clean
- `cargo clippy --workspace --all-targets -- -D warnings` clean (locally; `--all-features` exposes a pre-existing macOS env limitation with sz3-sys's `openmp` feature, which CI's Linux container handles natively)
- `cargo test --workspace` all green
- `cargo test -p tensogram --features remote,async` all green
- `cargo build --workspace` clean
- `mdbook build docs/` clean
- `ruff check` + `ruff format --check` clean (all of `python/tests/`)
- `pytest python/tests/` 628 passed, 2 skipped (inverse stubs)
- `pytest python/tensogram-xarray/tests/` 245 passed
- `pytest python/tensogram-zarr/tests/` 235 passed

## What this PR does NOT do

- Does not run any mutation tests at merge time.
- Does not flip `mutants-diff` to required-for-merge — that awaits one clean weekly run as observation.
- Does not modify any tensogram library source. New tests only; existing source untouched.
- Does not extend mutation testing to other-language bindings (Python / C++ / WASM / TS) — those rely on the existing golden-file byte-parity tests for cross-language correctness.

## How will you know if a weekly run finds a problem?

Three independent signals (`MUTATION_TESTING.md` §4.3):

1. **GitHub Issue auto-created** with `mutation-testing` + `triage` labels. Body has status table, full `missed.txt`, link to artifact. Repo watchers receive an email.
2. **Workflow run summary** (`\$GITHUB_STEP_SUMMARY`) — totals + first 50 surviving mutants visible directly on the run page.
3. **Red badge** on the Actions tab.

7-day triage SLA per surviving-mutant issue.

Docs Preview
https://sites.ecmwf.int/docs/tensogram/pull-requests/PR-114

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -679,3 +612,41 @@ jobs:
...
+        password: ${{ secrets.ECMWF_DOCKER_REGISTRY_ACCESS_TOKEN }}
+    if: github.event_name == 'pull_request'
+    timeout-minutes: 30
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0  # need base branch for --in-diff
+      - run: mkdir -p "$TMPDIR"
+      - run: cargo install cargo-mutants --version 27.0.0 --locked
+      - name: Mutate diff vs base
```
Comment (tmi) [inline .github/workflows/ci.yml:645]: I would suggest having these three steps as a single `make` action, possibly exposing parallelism level (the cargo mutant jobs; the tgm threads / cmake -j should stay at 1). The less "business logic" in github actions, the better

Comment (tmi) [reply .github/workflows/ci.yml:645]: *especially* since its shared with the cron action -- we want a guaranteed single impl of this

Comment (tlmquintino) [reply .github/workflows/ci.yml:645]: makes sense. lets do that.

File: .github/workflows/mutants-weekly.yml [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+  schedule:
+    # Saturday + Sunday 09:00 UTC = 10:00-11:00 CET/CEST.
+    # Daytime weekend — minimal contention on shared builder fleet.
+    - cron: "0 9 * * 6,0"
+  workflow_dispatch:
+    inputs:
+      shards:
+        description: "Number of shards (default 16; reduce for ad-hoc smaller runs)"
+        required: false
+        default: "16"
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply .github/workflows/mutants-weekly.yml]: Fixed in 5aff5f2 — removed the unused `workflow_dispatch.inputs.shards`. `workflow_dispatch: {}` now (manual trigger still works, just without a knob since the matrix is hard-coded `[1..16]`).

File: .github/workflows/mutants-weekly.yml [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+      # Wallclock for 16 shards becomes 16/2 = 8 batches × ~30 min
+      # = ~4 hr total weekend wallclock.
+      max-parallel: 2
+      matrix:
+        shard:
+          [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]
+    steps:
+      - uses: actions/checkout@v4
+      - run: mkdir -p "$TMPDIR"
+      - run: cargo install cargo-mutants --version 27.0.0 --locked
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply .github/workflows/mutants-weekly.yml]: Fixed in 5aff5f2 — added `env: TMPDIR: ${{ github.workspace }}/.tmp` at workflow level, matching the `ci.yml` convention. Both the per-shard job and the new `collate` job now `mkdir -p "$TMPDIR"` reliably.

File: docs/src/dev/mutation-testing.md [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+  example.
+- **Wire-format byte-equality.** Tensogram guarantees that the same
+  input produces bit-identical output across Rust, Python, C++, and
+  TypeScript. Golden-file tests enforce this, but mutation testing
+  reveals whether those golden files actually exercise every branch
+  in the encoder and decoder.
+
+The full rollout plan, including phasing, shard strategy, and triage
+SLAs, lives in [`MUTATION_TESTING.md`](../../../MUTATION_TESTING.md) at
+the repository root.
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply docs/src/dev/mutation-testing.md:34]: Fixed in 5aff5f2 — `MUTATION_TESTING.md` is now committed at repo root. All four cross-references (this file, `Cargo.toml:71`, `.cargo/mutants.toml:4`, `plans/TEST.md:53`) now resolve. The plan archives into `plans/DONE.md` once Phase 2 closes (per the convention used for other PLAN_*.md files).

File: Cargo.toml [hunk 5]
```diff
@@ -58,3 +58,17 @@ panic = "abort"
...
+# Profile used by `cargo mutants` for both the build and test phases.
+# Inherits from `release` (so test wallclock is 8x faster than dev) but
+# pins `codegen-units = 1` to serialise rustc's internal codegen
+# threads.  Without this pin, each rustc invocation in release mode
+# defaults to 16 codegen threads — and cargo's jobserver only counts
+# the rustc *process* as one slot, not its 16 internal threads.  With
+# `--jobserver-tasks 2` that's 2 × 16 = 32 simultaneous codegen threads
+# during peak build.  On a 12-core laptop this triggers brief peak-power
+# events that load average smooths over but a power supply does not.
+# See MUTATION_TESTING.md §3.0.1 for the full story.
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply Cargo.toml:71]: Fixed in 5aff5f2 — `MUTATION_TESTING.md` is now committed at repo root. Reference resolves.

File: plans/TEST.md [hunk 6]
```diff
@@ -28,11 +28,49 @@ Repo: ecmwf/tensogram
...
+**Tool:** `cargo-mutants` 27.0.0 (pinned). Rust workspace only — does not extend
+to Python, C++, TypeScript, or WASM bindings (those rely on golden-file
+byte-parity tests).
+
+**Regimes:**
+- **PR-time** (`--in-diff origin/main..HEAD`): non-blocking job in
+  `.github/workflows/ci.yml`; uploads `mutants.out/` as artifact.
+  Flip to required-for-merge after Phase-1 stabilises (see rollout plan).
+- **Weekly full sweep** (sharded 1/16..16/16, Sat+Sun 09:00 UTC): `.github/workflows/mutants-weekly.yml`,
+  weekdays 02:00 UTC. Surviving mutants → auto-issue tagged `mutation-testing`.
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply plans/TEST.md]: Fixed in 5aff5f2 — schedule line was a leftover from the nightly→weekly rename. Now reads "Saturday + Sunday 09:00 UTC" only (consistent with `mutants-weekly.yml`'s actual cron). Also switched the workflow reference to `make mutants-diff` (the new wrapper, see tmi's comment).

File: plans/TEST.md [hunk 7]
```diff
@@ -28,11 +28,49 @@ Repo: ecmwf/tensogram
...
+
+**Config:** `.cargo/mutants.toml` at repo root. Excludes: `remote.rs` (I/O-driven),
+`doctor/**`, `tensogram-cli/**`, benchmarks, examples, `tensogram-sz3-sys`.
+Rollout plan: `MUTATION_TESTING.md` (repo root). User-facing docs:
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply plans/TEST.md:53]: Fixed in 5aff5f2 — `MUTATION_TESTING.md` is now committed. Reference resolves.

File: .cargo/mutants.toml [hunk 8]
```diff
@@ -0,0 +1,197 @@
+# .cargo/mutants.toml
+#
+# See MUTATION_TESTING.md for the rollout plan.
+# See https://mutants.rs/ for cargo-mutants usage.
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply .cargo/mutants.toml:4]: Fixed in 5aff5f2 — `MUTATION_TESTING.md` is now committed. Header-comment reference resolves.

File: .github/workflows/mutants-weekly.yml [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+concurrency:
+  group: mutants-weekly
+  cancel-in-progress: false
+
+# Explicit least-privilege GITHUB_TOKEN permissions for this workflow.
+permissions:
+  contents: read
+  actions: read
+  issues: write
+
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply .github/workflows/mutants-weekly.yml:54]: Fixed in 5aff5f2 — was a real duplicate (two earlier Copilot Autofix commits each added a top-level `permissions:` block; the second's `issues: write` flavour overrode the first's read-only flavour per YAML semantics). Collapsed into one workflow-level read-only block; `issues: write` stays scoped to just the `open-issue-on-fail` job.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (tlmquintino) [reply .github/workflows/ci.yml:645]: Done in 5aff5f2 — three new Makefile targets:

- `make mutants-install` — idempotent `cargo install cargo-mutants --version 27.0.0 --locked`
- `make mutants-diff` — `git diff origin/main...HEAD > tmpfile && cargo mutants --in-diff tmpfile` (also fixes the live CI failure, which was caused by passing a git rev range to a flag that wants a file path)
- `make mutants-shard SHARD=N/M` — single-shard run (used by the weekly workflow's matrix)
- `make mutants` — full workspace sweep (long; for ad-hoc use)

`MUTANTS_JOBS` is the only tunable knob; the other six clamps (TENSOGRAM_THREADS=1, CMAKE_BUILD_PARALLEL_LEVEL=1, MAKEFLAGS=-j1, --jobserver-tasks tied to MUTANTS_JOBS, --no-shuffle, --timeout-multiplier 3.0) are pinned in a shared `MUTANTS_ENV` Make macro because they MUST stay clamped to prevent the nested-fan-out laptop power-down events documented in `docs/src/dev/mutation-testing.md` 'Concurrency'.

`ci.yml`'s mutants-diff job collapses to one `run: make mutants-diff MUTANTS_JOBS=2` step (was three steps); same shape in `mutants-weekly.yml`'s per-shard step.
